sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala

```diff
@@ -132,7 +132,9 @@ class HadoopTableReader(
     val deserializedHadoopRDD = hadoopRDD.mapPartitions { iter =>
       val hconf = broadcastedHadoopConf.value.value
       val deserializer = deserializerClass.getConstructor().newInstance()
-      deserializer.initialize(hconf, localTableDesc.getProperties)
+      DeserializerLock.synchronized {
```

Comment:
Is there any other way to guard against this besides a global lock? that could introduce contention. `deserializer` is a local object here; it accesses some non-thread-safe shared state?

Reply:
I don't think there will be a lot of contention. We call `deserializer.initialize` once for each partition.
The race is actually in `HCatRecordObjectInspectorFactory.getHCatRecordObjectInspector` which is called by `JsonSerDe#initialize` (the factory uses a `HashMap` instead of , e.g., a `ConcurrentHashMap` to maintain a cache). We want the same `ObjectInspector` to be returned by the factory cache and set in each `JsonSerDe` instance as its `ObjectInspector`.
Our objective is that there not be more than one task from each executor calling `JsonSerDe#initialize` at the same time.

Reply:
Fair enough. Is there any way to fix this in the Hadoop implementation? or is it already fixed but we have to deal with it here? 
Can we push the global lock down further rather than making callers know to use it?

Reply:
HIVE-15773 and HIVE-21752 offer two different fixes for the `HCatRecordObjectInspectorFactory` bug. HIVE-21752 has been merged in Hive 4.0.0, which is not released yet. Even when it is, not many users will be on Hive 4.0.0 for some time, so I think we need to deal with it here.

Reply:
On the question of other callers needing to use this global lock:
AFAIK, the only reported problem in Spark with the `HCatRecordObjectInspectorFactory` bug is in reading partitioned `JsonSerDe` tables. In `HadoopTableReader#makeRDDForTable`, we can even do without using the lock; we only really need to use the lock in `HadoopTableReader#makeRDDForPartitionedTable`.
From my search, there are two other places in Spark where we call `Deserializer#initialize`:
1. `HiveTableScanExec`, in initializing a `HadoopTableReader` instance, before `HadoopTableReader#makeRDDForTable` or `HadoopTableReader#makeRDDForPartitionedTable` even get called in `doExec`.
2. `HiveScriptIOSchema`.

I don't think we need to use the lock for 1.
I don't know about 2., but if no problem has been reported due to the race, we can also leave it alone.
In other words, I'm not proposing we guard against the race in the Hive bug everywhere, just in this known case.